### PR TITLE
feature(zjow): add new pipeline agent sac/ddpg/a2c

### PR DESCRIPTION
**Pull request by zjowowen**
Add new agent class for sac/ddpg/a2c algorithms.

### CODE REVIEW

File: ding/bonus/ppof.py [hunk 1]
```diff
@@ -206,7 +221,7 @@ def collect_data(
             task.use(offline_data_saver(save_data_path, data_type='hdf5'))
             task.run(max_step=1)
         logging.info(
-            f'PPOF collecting is finished, more than {n_sample} samples are collected and saved in `{save_data_path}`'
+            'PPOF collecting is finished, more than {n_sample} samples are collected and saved in `{save_data_path}`'
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:224]: why remove f

File: ding/bonus/ppof.py [hunk 2]
```diff
@@ -111,6 +125,7 @@ def __init__(
         self.policy = PPOFPolicy(self.cfg, model=model)
         if policy_state_dict is not None:
             self.policy.load_state_dict(policy_state_dict)
+        self.model_save_dir = os.path.join(self.exp_name, "model")
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:128]: `ckpt` maybe better

File: ding/bonus/ppof.py [hunk 3]
```diff
@@ -248,3 +265,11 @@ def _setup_env_manager(
             if context is not None:
                 manager_cfg.context = context
         return env_cls([partial(self.env.clone, caller) for _ in range(env_num)], manager_cfg)
+
+    @property
+    def best(self):
+        best_model_file_path = os.path.join(self.model_save_dir, "ckpt", "eval.pth.tar")
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:271]: why another ckpt here

File: ding/bonus/ppof.py [hunk 4]
```diff
@@ -248,3 +265,11 @@ def _setup_env_manager(
...
+        if os.path.exists(best_model_file_path):
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:272]: why load here, it is not the duty of this method

File: ding/bonus/ppof.py [hunk 5]
```diff
@@ -27,6 +28,17 @@ class TrainingReturn:
     wandb_url: str
 
 
+@dataclass
+class EvalReturn:
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:32]: can we use the same EvalReturn for different policies

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/bonus/ppof.py (hunk 1 above)
Comment (zjowowen) [reply ding/bonus/ppof.py]: fixed

File: ding/bonus/ppof.py (hunk 2 above)
Comment (zjowowen) [reply ding/bonus/ppof.py]: Yes, fixed.

File: ding/bonus/ppof.py (hunk 3 above)
Comment (zjowowen) [reply ding/bonus/ppof.py]: fixed

File: ding/bonus/ppof.py (hunk 5 above)
Comment (zjowowen) [reply ding/bonus/ppof.py]: fixed

File: ding/framework/middleware/functional/advantage_estimator.py [hunk 6]
```diff
@@ -79,3 +79,16 @@ def _estimator(ctx: "OnlineRLContext"):
...
     return _estimator
+
+
+def pg_estimator(policy: Policy) -> Callable:
+
+    def _estimator(ctx: "OnlineRLContext"):
+        train_data = []
+        for episode in ctx.episodes:
+            data = ttorch_collate(episode)
+            data = policy.get_train_sample(data)
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/advantage_estimator.py:90]: don't use policy method, add all the implementation here

File: ding/framework/middleware/functional/advantage_estimator.py [hunk 7]
```diff
@@ -79,3 +79,16 @@ def _estimator(ctx: "OnlineRLContext"):
         ctx.train_data = data
 
     return _estimator
+
+
+def pg_estimator(policy: Policy) -> Callable:
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/advantage_estimator.py:84]: rename to `mc_estimator`?

File: ding/utils/default_helper.py [hunk 8]
```diff
@@ -470,7 +470,7 @@ def split_data_generator(data: dict, split_size: int, shuffle: bool = True) -> d
     # assert len(set(length)) == 1, "data values must have the same length: {}".format(length)
     # if continuous action, data['logit'] is list of length 2
     length = length[0]
-    assert split_size >= 1 and split_size <= length, f'{split_size}_{length}'
```
Comment (PaParaZz1) [inline ding/utils/default_helper.py:473]: why remove this

File: ding/policy/pg.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 from typing import List, Dict, Any, Tuple, Union
 from collections import namedtuple
 import torch
+import treetensor as ttorch
```
Comment (PaParaZz1) [inline ding/policy/pg.py:4]: don't modify this file

File: ding/framework/middleware/functional/trainer.py [hunk 10]
```diff
@@ -29,19 +29,20 @@ def _train(ctx: Union["OnlineRLContext", "OfflineRLContext"]):
 
         if ctx.train_data is None:
             return
-        data = ctx.train_data
         train_output = policy.forward(ctx.train_data)
         if ctx.train_iter % cfg.policy.learn.learner.hook.log_show_after_iter == 0:
+            if isinstance(train_output, list):
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/trainer.py:34]: why add this

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/functional/collector.py [hunk 11]
```diff
@@ -77,6 +77,9 @@ def _inference(ctx: "OnlineRLContext"):
 
         obs = {i: obs[i] for i in range(get_shape0(obs))}  # TBD
         inference_output = policy.forward(obs, **ctx.collect_kwargs)
+        # for key, value in inference_output.items():
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:80]: remove these codes

File: ding/policy/a2c.py [hunk 12]
```diff
@@ -274,3 +293,6 @@ def _forward_eval(self, data: dict) -> dict:
 
     def _monitor_vars_learn(self) -> List[str]:
         return super()._monitor_vars_learn() + ['policy_loss', 'value_loss', 'entropy_loss', 'adv_abs_max', 'grad_norm']
+
+    def monitor_vars(self) -> List[str]:
```
Comment (PaParaZz1) [inline ding/policy/a2c.py:297]: why add this

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/functional/advantage_estimator.py [hunk 13]
```diff
@@ -28,6 +28,12 @@ def gae_estimator(cfg: EasyDict, policy: Policy, buffer_: Optional[Buffer] = Non
     """
 
     model = policy.get_attribute('model')
+    obs_shape = cfg['policy']['model']['obs_shape']
+    obs_shape = torch.Size(torch.tensor(obs_shape)) if isinstance(obs_shape, list) \
+        else torch.Size(torch.tensor(obs_shape).unsqueeze(0))
```
Comment (zjowowen) [inline ding/framework/middleware/functional/advantage_estimator.py:33]: check which policy need it.

File: ding/framework/middleware/functional/logger.py [hunk 14]
```diff
@@ -153,10 +152,17 @@ def wandb_online_logger(
         metric_list = ["q_value", "target q_value", "loss", "lr", "entropy", "target_q_value", "td_error"]
     # Initialize wandb with default settings
     # Settings can be covered by calling wandb.init() at the top of the script
-    if anonymous:
-        wandb.init(project=project_name, reinit=True, anonymous="must")
+    if not wandb_sweep:
```
Comment (zjowowen) [inline ding/framework/middleware/functional/logger.py:155]: delete "if not"

File: ding/framework/middleware/functional/logger.py [hunk 15]
```diff
@@ -197,7 +203,12 @@ def _plot(ctx: "OnlineRLContext"):
 
         if cfg.plot_logger:
             for metric in metric_list:
-                if metric in ctx.train_output[0]:
+                if isinstance(ctx.train_output, Dict) and metric in ctx.train_output:
```
Comment (zjowowen) [inline ding/framework/middleware/functional/logger.py:206]: unify format ctx.train_output.

File: ding/utils/default_helper.py [hunk 16]
```diff
@@ -461,7 +463,10 @@ def split_data_generator(data: dict, split_size: int, shuffle: bool = True) -> d
         elif k in ['prev_state', 'prev_actor_state', 'prev_critic_state']:
             length.append(len(v))
         elif isinstance(v, list) or isinstance(v, tuple):
-            length.append(get_shape0(v[0]))
+            if isinstance(v[0], str):
```
Comment (PaParaZz1) [inline ding/utils/default_helper.py:466]: add comments for this case

File: ding/model/common/utils.py [hunk 17]
```diff
@@ -1,4 +1,5 @@
 import torch
+import torch.nn as nn
```
Comment (PaParaZz1) [inline ding/model/common/utils.py:2]: why add this

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/c51.py [hunk 18]
```diff
@@ -266,3 +266,6 @@ def _get_train_sample(self, data: list) -> Union[None, List[Any]]:
         """
         data = get_nstep_return_data(data, self._nstep, gamma=self._gamma)
         return get_train_sample(data, self._unroll_len)
+
+    def monitor_vars(self) -> List[str]:
+        return ['cur_lr', 'total_loss', 'q_value', 'target_q_value']
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/policy/c51.py:271]: fixed

File: ding/framework/middleware/functional/advantage_estimator.py (hunk 13 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/advantage_estimator.py:33]: for PG continuous

File: ding/policy/a2c.py (hunk 12 above)
Comment (zjowowen) [reply ding/policy/a2c.py:297]: fixed

File: ding/utils/default_helper.py (hunk 16 above)
Comment (zjowowen) [reply ding/utils/default_helper.py:466]: fixed

File: ding/model/common/utils.py (hunk 17 above)
Comment (zjowowen) [reply ding/model/common/utils.py]: fixed.

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/envs/ding_env_manager.py [hunk 19]
```diff
@@ -0,0 +1,23 @@
+from .env_manager import BaseEnvManagerV2, SubprocessEnvManagerV2
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/envs/ding_env_manager.py]: fixed

File: ding/policy/pg.py [hunk 20]
```diff
@@ -169,16 +169,26 @@ def _get_train_sample(self, data: list) -> Union[None, List[Any]]:
...
-            data[i]['return'] = R
-        return get_train_sample(data, self._unroll_len)
+        if isinstance(data, list):
+            for i in reversed(range(len(data))):
+                R = self._gamma * R + data[i]['reward']
+                data[i]['return'] = R
+            return get_train_sample(data, self._unroll_len)
+        elif isinstance(data, ttorch.Tensor):
+            data_size = data['done'].shape[0]
+            data['return'] = ttorch.Tensor([0.0 for i in range(data_size)])
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/policy/pg.py]: fixed

File: ding/framework/middleware/functional/advantage_estimator.py (hunk 13 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/advantage_estimator.py:34]: fixed

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**